File: src/renderground/geometry/mesh.cpp
#include <cassert>

#include "geometry/mesh.h"
#include "math/wrap.h"
#include "math/constants.h"

namespace ground {

Mesh::Mesh(const Vector3* verts, int numVerts, const int* indices, int numIndices,
    const Vector2* texCoords, const Vector3* shadingNormals)
: vertices(numVerts), indices(numIndices)
{
    // TODO assert that indices is a multiple of two
    //      (sanity check level, i.e., not only in debug mode)
    const int numFaces = numIndices / 3;

    std::copy(verts, verts + numVerts, this->vertices.begin());
    std::copy(indices, indices + numIndices, this->indices.begin());

    // Compute surface normals and areas
    faceNormals.resize(numFaces);
    surfaceAreas.resize(numFaces);
    totalSurfaceArea = 0;
    for (int face = 0; face < numFaces; ++face) {
        auto& v1 = vertices[indices[face * 3 + 0]];
        auto& v2 = vertices[indices[face * 3 + 1]];
        auto& v3 = vertices[indices[face * 3 + 2]];

        // Compute the normal. Winding order is CCW always.
        Vector3 n = Cross(v2 - v1, v3 - v1);
        float len = Length(n);
        faceNormals[face] = n / len;
        surfaceAreas[face] = len * 0.5f;

        totalSurfaceArea += surfaceAreas[face];
    }

    triangleDistribution.Build(surfaceAreas.begin(), surfaceAreas.end());

    // Per-vertex texture coordinates
    this->textureCoordinates.resize(numVerts);
    if (!texCoords) { // not given: default to (0,0) everywhere
        std::fill(textureCoordinates.begin(), textureCoordinates.end(), Vector2{0, 0});
    } else {
        std::copy(texCoords, texCoords + numVerts, textureCoordinates.begin());
    }

    // Per-vertex shading normals
    this->shadingNormals.resize(numVerts);
    if (!shadingNormals) { // not given: default to face normals
        for (int face = 0; face < numFaces; ++face) {
            this->shadingNormals[indices[face * 3 + 0]] = faceNormals[face];
            this->shadingNormals[indices[face * 3 + 1]] = faceNormals[face];
            this->shadingNormals[indices[face * 3 + 2]] = faceNormals[face];
        }
    } else {
        std::copy(shadingNormals, shadingNormals + numVerts, this->shadingNormals.begin());
    }

    // TODO pre-normalize shading normals here? Can that be problematic with interpolation?
}

SurfacePoint Mesh::PrimarySampleToSurface(const Vector2& primarySample, float* jacobian) const {
    // TODO sanity check that both primary sample values
    //      are in [0,1]

    // Select a triangle proportionally to its surface area
    float selectionJacobian = 1.0f;
    unsigned int primId = triangleDistribution.TransformPrimarySample(
        primarySample.x, &selectionJacobian);

    // Remap the first PSS dimension to the selected triangle
    float lo = primId == 0 ? 0.0f : triangleDistribution.GetJacobian(primId - 1);
    float delta = selectionJacobian;
    float remapped = (primarySample.x - lo) * delta;
    assert(remapped >= 0.0f && remapped <= 1.0f);

    // Remap to a uniform distribution of barycentric coordinates
    float u = 0, v = 0;
    WrapToUniformTriangle(remapped, primarySample.y, u, v);
    Vector2 barycentricCoords { u, v };

    *jacobian = selectionJacobian / surfaceAreas[primId];
    CheckFloatEqual(*jacobian, 1.0f / totalSurfaceArea);

    return SurfacePoint {
        PointFromBarycentric(primId, barycentricCoords),
        faceNormals[primId],
        barycentricCoords,
        0xFFFFFFFF, // To be filled by the caller, we don't know our own Id
        primId,
    };
}

float Mesh::ComputePrimaryToSurfaceJacobian(const SurfacePoint& point) const {
    return 1.0f / totalSurfaceArea;
}

template<typename VertArray, typename IdxArray>
auto InterpolateOnTriangle(int primId, const Vector2& barycentric,
    const VertArray& vertexData, const IdxArray& indices)
{
    auto& v1 = vertexData[indices[primId * 3 + 0]];
    auto& v2 = vertexData[indices[primId * 3 + 1]];
    auto& v3 = vertexData[indices[primId * 3 + 2]];

    return barycentric.x * v2
         + barycentric.y * v3
         + (1 - barycentric.x - barycentric.y) * v1;
}

Vector3 Mesh::PointFromBarycentric(int primId, const Vector2& barycentric) const {
    return InterpolateOnTriangle(primId, barycentric, vertices, indices);
}

Vector2 Mesh::ComputeTextureCoordinates(int primId, const Vector2& barycentric) const {
    return InterpolateOnTriangle(primId, barycentric, textureCoordinates, indices);
}

Vector3 Mesh::ComputeShadingNormal(int primId, const Vector2& barycentric) const {
    // TODO this frequent call to Normalize could possibly be optimized away
    return Normalize(InterpolateOnTriangle(primId, barycentric, shadingNormals, indices));
}

} // namespace ground

File: src/SeeSharp/Core/Geometry/Mesh.cs
using SeeSharp.Core.Sampling;
using SeeSharp.Core.Shading.Materials;
using System.Diagnostics;
using System.Numerics;

namespace SeeSharp.Core.Geometry {

    public class Mesh {
        public Material Material;

        public Mesh(Vector3[] vertices, int[] indices, Vector3[] shadingNormals = null,
                    Vector2[] textureCoordinates = null) {
            Vertices = vertices;
            Indices = indices;

            Debug.Assert(indices.Length % 3 == 0, "Triangle mesh indices must be a multiple of three.");
            NumFaces = indices.Length / 3;
            NumVertices = vertices.Length;

            // Compute face normals and triangle areas
            FaceNormals = new Vector3[NumFaces];
            var surfaceAreas = new float[NumFaces];
            SurfaceArea = 0;
            for (int face = 0; face < NumFaces; ++face) {
                var v1 = vertices[indices[face * 3 + 0]];
                var v2 = vertices[indices[face * 3 + 1]];
                var v3 = vertices[indices[face * 3 + 2]];

                // Compute the normal. Winding order is CCW always.
                Vector3 n = Vector3.Cross(v2 - v1, v3 - v1);
                float len = n.Length();
                FaceNormals[face] = n / len;
                surfaceAreas[face] = len * 0.5f;

                SurfaceArea += surfaceAreas[face];
            }

            triangleDistribution = new Sampling.PiecewiseConstant(surfaceAreas);

            this.shadingNormals = shadingNormals;
            this.textureCoordinates = textureCoordinates;

            // Compute shading normals from face normals if not set
            if (this.shadingNormals == null) {
                this.shadingNormals = new Vector3[vertices.Length];
                for (int face = 0; face < NumFaces; ++face) {
                    this.shadingNormals[indices[face * 3 + 0]] = FaceNormals[face];
                    this.shadingNormals[indices[face * 3 + 1]] = FaceNormals[face];
                    this.shadingNormals[indices[face * 3 + 2]] = FaceNormals[face];
                }
            } else {
                // Ensure normalization
                for (int i = 0; i < this.shadingNormals.Length; ++i)
                    this.shadingNormals[i] = Vector3.Normalize(this.shadingNormals[i]);
            }
        }

        float ComputeErrorOffset(int faceIdx, Vector2 barycentricCoords) {
            var v1 = Vertices[Indices[faceIdx * 3 + 0]];
            var v2 = Vertices[Indices[faceIdx * 3 + 1]];
            var v3 = Vertices[Indices[faceIdx * 3 + 2]];

            Vector3 errorDiagonal = Vector3.Abs(barycentricCoords.X * v2)
                + Vector3.Abs(barycentricCoords.Y * v3)
                + Vector3.Abs((1 - barycentricCoords.X - barycentricCoords.Y) * v1);

            return errorDiagonal.Length() * 32.0f * 1.19209e-07f;
        }

        public SurfaceSample Sample(Vector2 primarySample) {
            var (faceIdx, newX) = triangleDistribution.Sample(primarySample.X);
            var barycentric = SampleWrap.ToUniformTriangle(new Vector2(newX, primarySample.Y));

            return new SurfaceSample {
                point = new SurfacePoint {
                    barycentricCoords = barycentric,
                    primId = (uint)faceIdx,
                    normal = FaceNormals[faceIdx],
                    position = ComputePosition(faceIdx, barycentric),
                    errorOffset = ComputeErrorOffset(faceIdx, barycentric),
                    mesh = this
                },
                pdf = 1.0f / SurfaceArea
            };
        }

        public float Pdf(SurfacePoint point) {
            return 1.0f / SurfaceArea;
        }

        public Vector2 ComputeTextureCoordinates(int faceIdx, Vector2 barycentric) {
            if (textureCoordinates == null)
                return new Vector2(0, 0);

            var v1 = textureCoordinates[Indices[faceIdx * 3 + 0]];
            var v2 = textureCoordinates[Indices[faceIdx * 3 + 1]];
            var v3 = textureCoordinates[Indices[faceIdx * 3 + 2]];

            return barycentric.X * v2
                +  barycentric.Y * v3
                + (1 - barycentric.X - barycentric.Y) * v1;
        }

        public Vector3 ComputeShadingNormal(int faceIdx, Vector2 barycentric) {
            var v1 = shadingNormals[Indices[faceIdx * 3 + 0]];
            var v2 = shadingNormals[Indices[faceIdx * 3 + 1]];
            var v3 = shadingNormals[Indices[faceIdx * 3 + 2]];

            return barycentric.X * v2
                +  barycentric.Y * v3
                + (1 - barycentric.X - barycentric.Y) * v1;
        }

        public Vector3 ComputePosition(int faceIdx, Vector2 barycentric) {
            var v1 = Vertices[Indices[faceIdx * 3 + 0]];
            var v2 = Vertices[Indices[faceIdx * 3 + 1]];
            var v3 = Vertices[Indices[faceIdx * 3 + 2]];

            return barycentric.X * v2
                +  barycentric.Y * v3
                + (1 - barycentric.X - barycentric.Y) * v1;
        }

        public Vector3[] Vertices;
        public int[] Indices;
        public Vector3[] FaceNormals;
        public float SurfaceArea;

        public int NumVertices { get; private set; }
        public int NumFaces { get; private set; }

        Sampling.PiecewiseConstant triangleDistribution;

        // per-vertex attributes
        Vector3[] shadingNormals;
        Vector2[] textureCoordinates;
    }
}
